kvserver: `TestCheckConsistencyInconsistent` failed, data race in `TestingSetRedactable`

```
=== RUN   TestCheckConsistencyInconsistent
    test_log_scope.go:79: test logs captured to: /home/erik/.cache/bazel/_bazel_erik/58b8816bb117734f0a0eaaef7fcef95b/sandbox/linux-sandbox/11283/execroot/com_github_cockroachdb_cockroach/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestCheckConsistencyInconsistent3070084177
    test_log_scope.go:80: use -show-logs to present logs inline
==================
WARNING: DATA RACE
Write at 0x00c0007d9b68 by goroutine 47:
  github.com/cockroachdb/cockroach/pkg/util/log.TestingSetRedactable.func1()
      github.com/cockroachdb/cockroach/pkg/util/log/redact.go:199 +0xe4
  github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestCheckConsistencyInconsistent()
      github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:477 +0x242b
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).WaitForFullReplication()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:1244 +0x4f7
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:370 +0xa24
  github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start()
      github.com/cockroachdb/cockroach/pkg/server/testserver.go:498 +0x109
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).startServer()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:501 +0x10a
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:327 +0x4b5
  github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start()
      github.com/cockroachdb/cockroach/pkg/server/testserver.go:498 +0x109
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).startServer()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:501 +0x10a
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:327 +0x4b5
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.StartTestCluster()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:196 +0xa8
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.NewTestCluster()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:292 +0xcec
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.NewTestCluster()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:292 +0xcec
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.StartTestCluster()
      github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:195 +0x8b
  github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestCheckConsistencyInconsistent()
      github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:336 +0x9e4
  testing.tRunner()
      GOROOT/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      GOROOT/src/testing/testing.go:1306 +0x47

Previous read at 0x00c0007d9b68 by goroutine 42:
  github.com/cockroachdb/cockroach/pkg/util/log.(*loggerT).outputLogEntry()
      github.com/cockroachdb/cockroach/pkg/util/log/clog.go:312 +0x7f5
  github.com/cockroachdb/cockroach/pkg/util/log.logfDepth()
      github.com/cockroachdb/cockroach/pkg/util/log/channels.go:60 +0x385
  github.com/cockroachdb/cockroach/pkg/util/log.Infof()
      github.com/cockroachdb/cockroach/bazel-out/k8-fastbuild/bin/pkg/util/log/log_channels_generated.go:263 +0x1f0
  github.com/cockroachdb/cockroach/pkg/util/admission.(*GrantCoordinator).CPULoad()
      github.com/cockroachdb/cockroach/pkg/util/admission/granter.go:956 +0x3e
  github.com/cockroachdb/cockroach/pkg/util/admission.(*GrantCoordinator).CPULoad-fm()
      github.com/cockroachdb/cockroach/pkg/util/admission/granter.go:951 +0x57
  github.com/cockroachdb/cockroach/pkg/util/goschedstats.(*schedStatsTicker).getStatsOnTick()
      github.com/cockroachdb/cockroach/pkg/util/goschedstats/runnable.go:236 +0x3fa
  github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0.func1()
      github.com/cockroachdb/cockroach/pkg/util/goschedstats/runnable.go:169 +0x1c4

Goroutine 47 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      GOROOT/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      GOROOT/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1504 +0x9d1
  github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestMain()
      github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/main_test.go:49 +0x105
  main.main()
      main/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/testmain.go:1435 +0x5c6

Goroutine 42 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/goschedstats.init.0()
      github.com/cockroachdb/cockroach/pkg/util/goschedstats/runnable.go:157 +0x2b
==================
I220525 12:44:18.543257 1 (gostd) testmain.go:1414  [-] 1  Test //pkg/kv/kvserver:kvserver_test exited with error code 66
```

Jira issue: CRDB-16183

I unskipped the test under race and stressed it:
```
# unskip test
dev test pkg/kv/kvserver -f=TestCheckConsistencyInconsistent --stress --race
```

This did not hit the data race mentioned above. However, after about 10m of stress, it failed with the following, more concerning error:
```
--- FAIL: TestCheckConsistencyInconsistent (518.47s)
    test_log_scope.go:161: test logs captured to: /tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestCheckConsistencyInconsistent646488916
    test_log_scope.go:79: use -show-logs to present logs inline
    consistency_queue_test.go:407:
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:407
        	Error:      	Not equal:
        	            	expected: []byte{0xf1, 0xab, 0xd1, 0x41, 0x75, 0xd1, 0x79, 0x30, 0xd5, 0x36, 0x8d, 0xaf, 0x52, 0xc0, 0xd5, 0x88, 0x73, 0xc5, 0xb3, 0x6f, 0x81, 0x21, 0x67, 0x4f, 0xca, 0xce, 0x1, 0x68, 0xc5, 0x84, 0x6f, 0x7d, 0x54, 0xe5, 0x7e, 0xaa, 0x2e, 0xf1, 0x33, 0xdf, 0x8f, 0x45, 0xa2, 0x4a, 0x1e, 0x77, 0x80, 0xfd, 0xf3, 0x25, 0x10, 0x16, 0xd5, 0x83, 0x57, 0x15, 0x3c, 0x1f, 0x7f, 0x53, 0x94, 0xf0, 0x25, 0x10}
        	            	actual  : []byte{0x1d, 0x77, 0xab, 0xd8, 0x5d, 0x3c, 0xe9, 0x36, 0xc2, 0xaa, 0xef, 0x40, 0x1c, 0xf2, 0xfb, 0x18, 0xcd, 0x55, 0x16, 0x71, 0xf2, 0xf4, 0xfe, 0x39, 0x66, 0xa1, 0xa2, 0xe6, 0xc1, 0x84, 0x77, 0xc0, 0xf8, 0x25, 0x11, 0x1e, 0xc0, 0xc5, 0x79, 0x53, 0x9d, 0xe2, 0x32, 0x6b, 0x73, 0xb5, 0x3, 0xe9, 0x29, 0x5, 0x25, 0x75, 0xe, 0xa4, 0x1d, 0xbd, 0x88, 0xf4, 0x6c, 0x53, 0xa9, 0x22, 0xef, 0xdc}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,6 +1,6 @@
        	            	 ([]uint8) (len=64) {
        	            	- 00000000  f1 ab d1 41 75 d1 79 30  d5 36 8d af 52 c0 d5 88  |...Au.y0.6..R...|
        	            	- 00000010  73 c5 b3 6f 81 21 67 4f  ca ce 01 68 c5 84 6f 7d  |s..o.!gO...h..o}|
        	            	- 00000020  54 e5 7e aa 2e f1 33 df  8f 45 a2 4a 1e 77 80 fd  |T.~...3..E.J.w..|
        	            	- 00000030  f3 25 10 16 d5 83 57 15  3c 1f 7f 53 94 f0 25 10  |.%....W.<..S..%.|
        	            	+ 00000000  1d 77 ab d8 5d 3c e9 36  c2 aa ef 40 1c f2 fb 18  |.w..]<.6...@....|
        	            	+ 00000010  cd 55 16 71 f2 f4 fe 39  66 a1 a2 e6 c1 84 77 c0  |.U.q...9f.....w.|
        	            	+ 00000020  f8 25 11 1e c0 c5 79 53  9d e2 32 6b 73 b5 03 e9  |.%....yS..2ks...|
        	            	+ 00000030  29 05 25 75 0e a4 1d bd  88 f4 6c 53 a9 22 ef dc  |).%u......lS."..|
        	            	 }
        	Test:       	TestCheckConsistencyInconsistent
```

Another error:

```
--- FAIL: TestCheckConsistencyInconsistent (696.45s)
    test_log_scope.go:161: test logs captured to: /Users/pavel/go/src/github.com/cockroachdb/cockroach/tmp/_tmp/577bfc8ed46850d1aaafa923fa37ce78/logTestCheckConsistencyInconsistent994443092
    test_log_scope.go:79: use -show-logs to present logs inline
    consistency_queue_test.go:313: 
                Error Trace:    github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:313
                                                        github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:358
                Error:          Received unexpected error:
                                could not collect checksum from any replica
                                (1)
                                  | (opaque error wrapper)
                                  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
                                  | reportable 0:
                                  |
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).runConsistencyCheck
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_consistency.go:376
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).checkConsistencyImpl
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_consistency.go:96
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).CheckConsistency
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_consistency.go:84
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeAdminBatch
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:967
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_send.go:188
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store_send.go:206
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:203
                                  | github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal
                                  |     github.com/cockroachdb/cockroach/pkg/server/node.go:1169
                                  | github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch
                                  |     github.com/cockroachdb/cockroach/pkg/server/node.go:1243
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvpb._Internal_Batch_Handler.func1
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvpb/bazel-out/darwin_arm64-fastbuild/bin/pkg/kv/kvpb/kvpb_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvpb/api.pb.go:9975
                                  | github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.ServerInterceptor.func1
                                  |     github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:96
                                  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:1162
                                  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func3
                                  |     github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:268
                                  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:1165
                                  | github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor
                                  |     github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:92
                                  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:1165
                                  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1.1
                                  |     github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:235
                                  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr
                                  |     github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:322
                                  | github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func1
                                  |     github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:233
                                  | google.golang.org/grpc.chainUnaryInterceptors.func1.1
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:1165
                                  | google.golang.org/grpc.chainUnaryInterceptors.func1
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:1167
                                  | github.com/cockroachdb/cockroach/pkg/kv/kvpb._Internal_Batch_Handler
                                  |     github.com/cockroachdb/cockroach/pkg/kv/kvpb/bazel-out/darwin_arm64-fastbuild/bin/pkg/kv/kvpb/kvpb_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvpb/api.pb.go:9977
                                  | google.golang.org/grpc.(*Server).processUnaryRPC
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:1340
                                  | google.golang.org/grpc.(*Server).handleStream
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:1713
                                  | google.golang.org/grpc.(*Server).serveStreams.func1.2
                                  |     google.golang.org/grpc/external/org_golang_google_grpc/server.go:965
                                  | runtime.goexit
                                  |     src/runtime/asm_arm64.s:1172
                                Wraps: (2) could not collect checksum from any replica
                                Error types: (1) *errbase.opaqueWrapper (2) *errutil.leafError
                Test:           TestCheckConsistencyInconsistent
    panic.go:540: -- test log scope end --
FAIL
I230321 09:18:54.842662 1 (gostd) testmain.go:1494  [-] 1  Test //pkg/kv/kvserver:kvserver_test exited with error code 1


ERROR: exit status 1

1 runs completed, 1 failures, over 11m43s
context canceled

```

Here is what happened during one of the test failures:
```
--- FAIL: TestCheckConsistencyInconsistent (132.90s)
...
    consistency_queue_test.go:410: 
                Error Trace:    github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/consistency_queue_test.go:410
                Error:          Not equal: 
                                expected: []byte{0x55, 0x7a, 0x1, 0x36, 0x95, 0xb1, 0x3, 0x8a, 0xa6, 0x4, 0xd4, 0x50, 0x63, 0x78, 0xf0, 0x57, 0x43, 0x20, 0x5, 0xb2, 0x6d, 0xf0, 0x47, 0x91, 0x36, 0x9c, 0x87, 0xd8, 0xa6, 0xff, 0x72, 0xf6, 0xaa, 0xf1, 0xf2, 0xd1, 0xc8, 0x24, 0x9e, 0xc1, 0x26, 0xa6, 0xe4, 0xae, 0x47, 0xb9, 0x9b, 0xb5, 0xd1, 0x15, 0xa5, 0x42, 0x9e, 0xca, 0xfa, 0x4c, 0x36, 0xc6, 0xf0, 0xce, 0x1f, 0xd5, 0x2e, 0x62}
                                actual  : []byte{0x5e, 0x10, 0x5b, 0x87, 0xb6, 0x42, 0x3, 0x1f, 0xdf, 0x88, 0x73, 0x55, 0x2f, 0x4b, 0xfd, 0x7a, 0x2c, 0xcb, 0x49, 0x55, 0xa6, 0x74, 0x95, 0x1c, 0xe5, 0x90, 0x79, 0xc7, 0x54, 0x8f, 0x51, 0xa1, 0xbe, 0x83, 0xf2, 0x3a, 0xcd, 0xae, 0x1e, 0x4, 0x68, 0x2c, 0x9f, 0x34, 0x8c, 0xda, 0xad, 0xd7, 0xd4, 0x84, 0x22, 0x9f, 0xe, 0x7e, 0xc8, 0x9, 0x53, 0x7, 0xff, 0xaa, 0xda, 0xbd, 0xd, 0x42}
...
```

Node `n2` got intentionally corrupted, and the first phase of the consistency check failed (nodes 1 and 3 agree, 2 is corrupted):
```
E230321 09:30:19.389914 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 560
... 560 +(n1,s1):1: checksum 0346578bad85b436b1d0f519907ffb46613982c18e527b425ae8786e034628c800aa089c55346b3b4dc2d029437c6a7140051ba1bd58368ea59c049682f4f524
... 560 +- stats: contains_estimates:0 last_update_nanos:1679391017206907000 intent_age:0 gc_bytes_age:0 live_bytes:40 live_count:2 key_bytes:28 key_count:2 val_bytes:12 val_count:2 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:578 sys_count:7 abort_span_bytes:0 
... 560 +- stats.Sub(recomputation): last_update_nanos:1679391017206907000 
... 560 +(n3,s3):3: checksum 0346578bad85b436b1d0f519907ffb46613982c18e527b425ae8786e034628c800aa089c55346b3b4dc2d029437c6a7140051ba1bd58368ea59c049682f4f524
... 560 +- stats: contains_estimates:0 last_update_nanos:1679391017206907000 intent_age:0 gc_bytes_age:0 live_bytes:40 live_count:2 key_bytes:28 key_count:2 val_bytes:12 val_count:2 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:578 sys_count:7 abort_span_bytes:0 
... 560 +- stats.Sub(recomputation): last_update_nanos:1679391017206907000 
... 560 +(n2,s2):2: checksum fa276506de3cdeb9e2518c2272579e9f03101eb5627a70a5c8fc81c3caf7b76cb7d3d49e085a8c481015af8a8804e2831ce0ed9c060677a619d26bc010d0ad6b [minority]
... 560 +- stats: contains_estimates:0 last_update_nanos:1679391017206907000 intent_age:0 gc_bytes_age:0 live_bytes:40 live_count:2 key_bytes:28 key_count:2 val_bytes:12 val_count:2 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:578 sys_count:7 abort_span_bytes:0 
... 560 +- stats.Sub(recomputation): last_update_nanos:1679391017206907000 live_bytes:-20 live_count:-1 key_bytes:-14 key_count:-1 val_bytes:-6 val_count:-1 
... 561  consistency check failed; fetching details and shutting down minority (n2,s2):2
```

Shortly, all 3 nodes got a second phase `CheckConsistency` command which instructs to grab storage checkpoints at the time of application, and to kill node 2:

```
W230321 09:30:19.659257 2275 kv/kvserver/replica_consistency.go:670 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,raft] 563  creating checkpoint ‹r5_at_41› with spans [‹/Local/RangeID/{4""-7""}› ‹/Local/Range/{System/tsd-Table/3}› ‹/Local/Lock/Intent/Local/Range/{System/tsd-Table/3}› ‹/Local/Lock/Intent/{System/tsd-Table/3}› ‹/{System/tsd-Table/3}›]
W230321 09:30:19.738066 489 kv/kvserver/replica_consistency.go:670 ⋮ [T1,n1,s1,r5/1:‹/{Systemtse-Table/0}›,raft] 564  creating checkpoint ‹r5_at_41› with spans [‹/Local/RangeID/{4""-7""}› ‹/Local/Range/{System/tsd-Table/3}› ‹/Local/Lock/Intent/Local/Range/{System/tsd-Table/3}› ‹/Local/Lock/Intent/{System/tsd-Table/3}› ‹/{System/tsd-Table/3}›]
W230321 09:30:19.883564 3298 kv/kvserver/replica_consistency.go:670 ⋮ [T1,n3,s3,r5/3:‹/{Systemtse-Table/0}›,raft] 565  creating checkpoint ‹r5_at_41› with spans [‹/Local/RangeID/{4""-7""}› ‹/Local/Range/{System/tsd-Table/3}› ‹/Local/Lock/Intent/Local/Range/{System/tsd-Table/3}› ‹/Local/Lock/Intent/{System/tsd-Table/3}› ‹/{System/tsd-Table/3}›]
```

Nodes 2 and 3 quickly succeeded in creating the checkpoint @ log index 41:

```
W230321 09:30:20.706666 2275 kv/kvserver/replica_consistency.go:674 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,raft] 572  created checkpoint ‹auxiliary/checkpoints/r5_at_41›
W230321 09:30:20.811606 3298 kv/kvserver/replica_consistency.go:674 ⋮ [T1,n3,s3,r5/3:‹/{Systemtse-Table/0}›,raft] 574  created checkpoint ‹auxiliary/checkpoints/r5_at_41›
```

5+eps seconds later, the leaseholder ended the second phase of the consistency check:

```
E230321 09:30:26.045762 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 596  
E230321 09:30:26.045762 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 596 +(n2,s2):2: checksum 3931acdb478bf3a29e62b7c7aaa0ad983475b58c1e74be639d681dbade43ed405e30e97ab8f6678c3859df71bc613562e45eeea43b8cf3ed1e534dc30a481cdc
E230321 09:30:26.045762 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 596 +- stats: contains_estimates:0 last_update_nanos:1679391017206907000 intent_age:0 gc_bytes_age:0 live_bytes:40 live_count:2 key_bytes:28 key_count:2 val_bytes:12 val_count:2 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:578 sys_count:7 abort_span_bytes:0 
E230321 09:30:26.045762 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 596 +- stats.Sub(recomputation): last_update_nanos:1679391017206907000 live_bytes:-20 live_count:-1 key_bytes:-14 key_count:-1 val_bytes:-6 val_count:-1 
E230321 09:30:26.045762 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 596 +(n3,s3):3: checksum 5e105b87b642031fdf8873552f4bfd7a2ccb4955a67
4951ce59079c7548f51a1be83f23acdae1e04682c9f348cdaadd7d484229f0e7ec8095307ffaadabd0d42 [minority]
E230321 09:30:26.045762 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 596 +- stats: contains_estimates:0 last_update_nanos:1679391017206907000 intent_age:0 gc_bytes_age:0 live_bytes:40 live_count:2 key_bytes:28 key_count:2 val_bytes:12 val_count:2 intent_bytes:0 intent_count:0 separated_intent_count:0 range_key_count:0 range_key_bytes:0 range_val_count:0 range_val_bytes:0 sys_bytes:578 sys_count:7 abort_span_bytes:0 
E230321 09:30:26.045762 19038 kv/kvserver/replica_consistency.go:151 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 596 +- stats.Sub(recomputation): last_update_nanos:1679391017206907000 
E230321 09:30:26.086872 19038 kv/kvserver/replica_consistency.go:247 ⋮ [T1,n2,s2,r5/2:‹/{Systemtse-Table/0}›,*kvpb.CheckConsistencyRequest] 597  consistency check failed
```

Notably, only nodes 2 and 3 returned a checksum. Node 1 [timed out at 5 seconds](https://github.com/cockroachdb/cockroach/blob/c3b4ec97bae999185baea4cc1683179bbf8eadbb/pkg/kv/kvserver/replica_consistency.go#L413-L416) and returned an error.

Only 10 more seconds later, node 1 finally reported that it created a checksum: 

```
W230321 09:30:36.510150 489 kv/kvserver/replica_consistency.go:674 ⋮ [T1,n1,s1,r5/1:‹/{Systemtse-Table/0}›,raft] 647  created checkpoint ‹auxiliary/checkpoints/r5_at_41›
```

And exactly [5 seconds later it gave up](https://github.com/cockroachdb/cockroach/blob/c3b4ec97bae999185baea4cc1683179bbf8eadbb/pkg/kv/kvserver/replica_consistency.go#L704) on keeping the result of the checksum computation because the result collection request never showed up (we know that it timed out above):

```
E230321 09:30:41.521566 20700 kv/kvserver/replica_consistency.go:718 ⋮ [T1,n1,s1,r5/1:‹/{Systemtse-Table/0}›] 662  checksum collection did not join: ‹operation "kvserver.Replica: computing checksum" timed out after 5.01s (given timeout 5s)›: context deadline exceeded
```

So what's probably happening is:
- [runConsistency](https://github.com/cockroachdb/cockroach/blob/c3b4ec97bae999185baea4cc1683179bbf8eadbb/pkg/kv/kvserver/consistency_queue_test.go#L359-L364) check returns early when only nodes 2 and 3 have created the storage checkpoint.
- Node 1 haven't created the checkpoint, but has started doing so.
- Node 2 "fatals" (mocked out in the test) shortly after the check is completed.
- Node 1 is still creating its checkpoint, but has probably created the directory by now.
- Hence, the test assumes that the [checkpoint has been created](https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/consistency_queue_test.go#L374-L375).
- And [proceeds](https://github.com/cockroachdb/cockroach/blob/master/pkg/kv/kvserver/consistency_queue_test.go#L384-L408) to open it and compute the checksum of the range.

The problem is that on the node 1 the checkpoint directory isn't complete yet. Even the `RdbCheckpoints` metric check doesn't save it, because it's [updated](https://github.com/cockroachdb/cockroach/blob/c3b4ec97bae999185baea4cc1683179bbf8eadbb/pkg/kv/kvserver/store.go#L3118-L3122) asynchronously and on the basis of the number of directories in the checkpoints folder.

--- a/pkg/kv/kvserver/consistency_queue_test.go
+++ b/pkg/kv/kvserver/consistency_queue_test.go
@@ -348,12 +348,12 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 		assert.Empty(t, onDiskCheckpointPaths(i))
 	}
 
-	// Write some arbitrary data only to store 1. Inconsistent key "e"!
-	store1 := tc.GetFirstStoreFromServer(t, 1)
+	// Write some arbitrary data only to store on n2. Inconsistent key "e"!
+	s2 := tc.GetFirstStoreFromServer(t, 1)
 	var val roachpb.Value
 	val.SetInt(42)
 	// Put an inconsistent key "e" to s2, and have s1 and s3 still agree.
-	require.NoError(t, storage.MVCCPut(context.Background(), store1.TODOEngine(), nil,
+	require.NoError(t, storage.MVCCPut(context.Background(), s2.TODOEngine(), nil,
 		roachpb.Key("e"), tc.Server(0).Clock().Now(), hlc.ClockTimestamp{}, val, nil))
 
 	// Run consistency check again, this time it should find something.
@@ -369,6 +369,24 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 	assert.Contains(t, resp.Result[0].Detail, `[minority]`)
 	assert.Contains(t, resp.Result[0].Detail, `stats`)
 
+	// Make sure that all the stores started creating a checkpoint. The metric
+	// measures the number of checkpoint directories, but a directory can
+	// represent an incomplete checkpoint that is still being populated.
+	for i := 0; i < numStores; i++ {
+		metric := tc.GetFirstStoreFromServer(t, i).Metrics().RdbCheckpoints
+		testutils.SucceedsSoon(t, func() error {
+			if got, want := metric.Value(), int64(1); got != want {
+				return errors.Errorf("%s is %d, want %d", metric.Name, got, want)
+			}
+			return nil
+		})
+	}
+	// As discussed in https://github.com/cockroachdb/cockroach/issues/81819, it
+	// is possible that the check completes while there are still checkpoints in
+	// flight. Waiting for the server termination makes sure that checkpoints are
+	// fully created.
+	tc.Stopper().Stop(context.Background())
+
 	// Checkpoints should have been created on all stores.
 	hashes := make([][]byte, numStores)
 	for i := 0; i < numStores; i++ {
@@ -377,14 +395,6 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 		t.Logf("found a checkpoint at %s", cps[0])
 		// The checkpoint must have been finalized.
 		require.False(t, strings.HasSuffix(cps[0], "_pending"))
-
-		metric := tc.GetFirstStoreFromServer(t, i).Metrics().RdbCheckpoints
-		testutils.SucceedsSoon(t, func() error {
-			if got, want := metric.Value(), int64(1); got != want {
-				return errors.Errorf("%s is %d, want %d", metric.Name, got, want)
-			}
-			return nil
-		})
 
 		// Create a new store on top of checkpoint location inside existing in-mem
 		// VFS to verify its contents.
@@ -417,8 +427,8 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 	assert.Equal(t, hashes[0], hashes[2])    // s1 and s3 agree
 	assert.NotEqual(t, hashes[0], hashes[1]) // s2 diverged
 
-	// A death rattle should have been written on s2 (store index 1).
-	eng := store1.TODOEngine()
+	// A death rattle should have been written on s2.
+	eng := s2.TODOEngine()
 	f, err := eng.Open(base.PreventedStartupFile(eng.GetAuxiliaryDir()))
 	require.NoError(t, err)
 	b, err := io.ReadAll(f)

--- a/pkg/kv/kvserver/consistency_queue_test.go
+++ b/pkg/kv/kvserver/consistency_queue_test.go
@@ -35,7 +35,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -243,14 +242,6 @@ func TestCheckConsistencyReplay(t *testing.T) {
 func TestCheckConsistencyInconsistent(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-
-	// TODO(pavelkalinnikov): not if we remove TestingSetRedactable below?
-	skip.UnderRaceWithIssue(t, 81819, "slow test, and TestingSetRedactable triggers race detector")
-
-	// This test prints a consistency checker diff, so it's
-	// good to make sure we're overly redacting said diff.
-	// TODO(pavelkalinnikov): remove this since we don't print diffs anymore?
-	defer log.TestingSetRedactable(true)()
 
 	// Test expects simple MVCC value encoding.
 	storage.DisableMetamorphicSimpleValueEncoding(t)

